cnf-tests/testsuites/e2esuite/nmstate/nmstate_sriov.go

```diff
@@ -0,0 +1,200 @@
+					},
+					DesiredState: nmstateshared.State{
+						Raw: nmstateshared.RawState(fmt.Sprintf(nmstateDesiredStateTemplateStaticIP, testDevice.Name)),
+					},
+				},
+			}
+			err = client.Client.Create(context.Background(), &nodeNetworkConfigPolicy)
+			Expect(err).ToNot(HaveOccurred())
+			DeferCleanup(client.Client.Delete, context.Background(), &nodeNetworkConfigPolicy)
+
+			waitForNMStatePolicyToBeStable(&nodeNetworkConfigPolicy)
+
+			By("Configuring VF on device via SriovNetworkNodePolicy")
+			sriovNetworkNodePolicy, err := sriovnetwork.CreateSriovPolicy(sriovclient, "test-nmstate-", namespaces.SRIOVOperator, testDevice.Name, node, 10, "testnmstateresource", "netdevice")
+			Expect(err).ToNot(HaveOccurred())
+			DeferCleanup(sriovclient.Delete, context.Background(), sriovNetworkNodePolicy)
+
+			Eventually(func() int64 {
+				testedNode, err := sriovclient.Nodes().Get(context.Background(), node, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				resNum := testedNode.Status.Allocatable[corev1.ResourceName("openshift.io/testnmstateresource")]
+				capacity, _ := resNum.AsInt64()
+				return capacity
+			}, 10*time.Minute, time.Second).Should(Equal(int64(10)))
+		})
+	})
+})
+
+func findNonPrimarySriovDevicesAndNode(n *cluster.EnabledNodes) (string, []*sriovv1.InterfaceExt, error) {
+	errs := []error{}
+
+	retNode := ""
+	retDevices := []*sriovv1.InterfaceExt{}
+
+	for _, node := range n.Nodes {
+		devices, err := n.FindSriovDevices(node)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		mainDeviceName, err := getPrimaryNICForNode(node)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		nonPrimaryDevices := []*sriovv1.InterfaceExt{}
+		for _, device := range devices {
+			if device.Name == mainDeviceName {
+				continue
+			}
+			nonPrimaryDevices = append(nonPrimaryDevices, device)
+		}
+
+		if len(nonPrimaryDevices) > len(retDevices) {
+			retNode = node
+			retDevices = nonPrimaryDevices
+		}
+	}
+
+	if len(retDevices) == 0 {
+		return "", nil, fmt.Errorf("can't find any SR-IOV devices in cluster's nodes: %w", errors.Join(errs...))
+	}
+
+	return retNode, retDevices, nil
+}
+
+func getPrimaryNICForNode(node string) (string, error) {
+	nodeObj, err := client.Client.Nodes().Get(context.Background(), node, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	buf, err := nodes.ExecCommandOnMachineConfigDaemon(client.Client, nodeObj,
+		[]string{"/bin/bash", "-c", "chroot /rootfs /usr/bin/ovs-vsctl list-ports br-ex | /usr/bin/grep -v patch"})
+	if err != nil {
+		return "", err
+	}
+
+	exp, err := regexp.Compile("\r\n")
+	if err != nil {
+		return "", err
+	}
+
+	return exp.ReplaceAllString(string(buf), ""), nil
+}
+
+func waitForNMStatePolicyToBeStable(nmstatePolicy *nmstatev1.NodeNetworkConfigurationPolicy) {
+	key := runtimeclient.ObjectKey{Name: nmstatePolicy.Name}
+	Eventually(func(g Gomega) {
+		err := client.Client.Get(context.Background(), key, nmstatePolicy)
```

Comment:
the nmstatePolicy is using the arg var, which  creates confusion about the scope of the variables. Maybe have a new one.

Help functions imo are easier to read/be refactored if they are not having many assertions. I always suggest the assertion to be in the main block and have function that return error/true.

Reply:
>the nmstatePolicy is using the arg var, which creates confusion about the scope of the variables. Maybe have a new one.

good point. And since the function only use the name of the policy, I refactor it to pass a string only.

> Help functions imo are easier to read/be refactored if they are not having many assertions. I always suggest the assertion to be in the main block and have function that return error/true.

I'm not 100% with it: helper functions must be short (<20 lines, as a rule of thumb) and that's where they get their readability. On the other side, the call site needs to be way more readable, as it contains the complicated test logic. With the current implementation, the call site reads:

```
waitForNMStatePolicyToBeStable(nodeNetworkConfigPolicy.Name)
```

which reads straightforward to me.
With functions like these, we might lose reusability; a function that waits for different conditions is almost copy/paste, but that's another story.